### PR TITLE
Swift 5.0

### PR DESCRIPTION
**Pull request by norio-nomura**

### CODE REVIEW

File: Source/SourceKittenFramework/SwiftDeclarationAttributeKind.swift [hunk 1]
```diff
@@ -95,4 +95,21 @@ public enum SwiftDeclarationAttributeKind: String {
...
+    case _dynamicReplacement = "source.decl.attribute._dynamicReplacement"
+    case _effects = "source.decl.attribute._effects"
+    case _hasInitialValue = "source.decl.attribute._hasInitialValue"
+    case _hasStorage = "source.decl.attribute._hasStorage"
+    case _nonoverride = "source.decl.attribute._nonoverride"
+    case _private = "source.decl.attribute._private"
+    case _show_in_interface = "source.decl.attribute._show_in_interface"
+    case dynamicCallable = "source.decl.attribute.dynamicCallable"
+    // lower case variants are already exists
+    case gkInspectable = "source.decl.attribute.GKInspectable"
```
Comment (norio-nomura) [inline Source/SourceKittenFramework/SwiftDeclarationAttributeKind.swift]: The lowercase versions already exists:
```swift
    case gkinspectable = "source.decl.attribute.gkinspectable"
    case ibaction = "source.decl.attribute.ibaction"
    case iboutlet = "source.decl.attribute.iboutlet"
```
But I added camel case versions separately.
I have not confirmed whether SourceKit returns the camel case version.

File: Cartfile [hunk 2]
```diff
@@ -1,2 +1,2 @@
 github "drmohundro/SWXMLHash" ~> 4.7
-github "jpsim/Yams" ~> 1.0
+github "jpsim/Yams" "master"
```
Comment (jpsim) [inline Cartfile]: We should cut Yams 2.0 before merging this.

File: Cartfile.private [hunk 3]
```diff
@@ -1,2 +1,2 @@
-github "Carthage/Commandant" ~> 0.15
+github "Carthage/Commandant" "master"
```
Comment (jpsim) [inline Cartfile.private]: Similar to Yams, we'll need a new Commandant release before merging this.

File: Tests/SourceKittenFrameworkTests/StringTests.swift [hunk 4]
```diff
@@ -191,7 +191,7 @@ class StringTests: XCTestCase {
                 }
             }
             """
-        XCTAssertEqual(string.bridge().byteRangeToNSRange(start: 115, length: 42), NSRange(location: 115, length: 38))
+        XCTAssertEqual(string.bridge().byteRangeToNSRange(start: 115, length: 41), NSRange(location: 115, length: 38))
```
Comment (jpsim) [inline Tests/SourceKittenFrameworkTests/StringTests.swift]: 😱

File: azure-pipelines.yml [hunk 5]
```diff
@@ -21,11 +27,13 @@ jobs:
         mkdir -p build/reports/
         swiftlint lint --strict --reporter junit > build/reports/swiftlint.xml
       displayName: swiftlint
+      condition: contains(variables['containerImage'], 'swiftlint')
```
Comment (jpsim) [inline azure-pipelines.yml]: Nice! 😍

**Comment by norio-nomura** [comment]
There is a difference in the output of `DocInfoTests.testDocInfoRequest()` between Xcode 10.2 and `swift-5.0-RELEASE` for macOS downloaded from https://swift.org/download/. 🤔
The latter produces result same with `LinuxDocInfo@swift-5.0.json` added in d439f64.

**Comment by norio-nomura** [comment]
> There is a difference in the output of `DocInfoTests.testDocInfoRequest()` between Xcode 10.2 and `swift-5.0-RELEASE` for macOS downloaded from https://swift.org/download/. 🤔

https://bugs.swift.org/browse/SR-10190

**Comment by jpsim** [comment]
I'd like to merge https://github.com/jpsim/Yams/pull/181 so that we can have a stable release version to point to instead of `master`, which SwiftPM doesn't handle as well.

**Comment by norio-nomura** [comment]
Rebased and force pushed.
It still needs to fix failing `ModuleTests.testCommandantResultDocs()`.

**Comment by johnfairh** [comment]
Oops sorry about the `String.Iterator` thing, didn't retest that branch on 10.2.

File: Cartfile.private (hunk 3 above)
Comment (jpsim) [reply Cartfile.private]: Pending in https://github.com/Carthage/Commandant/pull/148

File: Source/SourceKittenFramework/XcodeBuildSetting.swift [hunk 6]
```diff
@@ -12,86 +12,11 @@ struct XcodeBuildSetting: Codable {
...
     let buildSettings: [String: String]
     /// The target name.
     let target: String
 
     subscript(dynamicMember member: String) -> String? {
         return buildSettings[member]
     }
-
-#if os(Linux) && !swift(>=4.2.1)
-    init(from decoder: Decoder) throws {
```
Comment (jpsim) [inline Source/SourceKittenFramework/XcodeBuildSetting.swift]: Why remove this?

File: Tests/SourceKittenFrameworkTests/SourceKitTests.swift [hunk 7]
```diff
@@ -12,13 +12,30 @@ import XCTest
...
     let task = Process()
-    task.launchPath = executable
     task.arguments = arguments
 
     let pipe = Pipe()
     task.standardOutput = pipe
 
-    task.launch()
+    do {
+    #if canImport(Darwin)
```
Comment (jpsim) [inline Tests/SourceKittenFrameworkTests/SourceKitTests.swift]: This pattern is repeated enough times to be worth a small helper. If it's internal to SourceKittenFramework we can still use it from tests by using `@testable import SourceKittenFramework` without polluting our public API.

File: Tests/SourceKittenFrameworkTests/SourceKitTests.swift [hunk 8]
```diff
@@ -54,9 +71,7 @@ private let sourcekitStrings: [String] = {
 }()
 
 private func sourcekitStrings(startingWith pattern: String) -> Set<String> {
-    return Set(sourcekitStrings.filter { string in
-        return string.range(of: pattern)?.lowerBound == string.startIndex
-    })
+    return Set(sourcekitStrings.filter { $0.hasPrefix(pattern) })
```
Comment (jpsim) [inline Tests/SourceKittenFrameworkTests/SourceKitTests.swift]: Nice

File: Tests/SourceKittenFrameworkTests/SourceKitTests.swift [hunk 9]
```diff
@@ -187,87 +168,34 @@ class SourceKitTests: XCTestCase {
...
+        expected.subtract([
+            .__raw_doc_comment, .__setter_access, ._borrowed, ._dynamicReplacement, ._effects, ._hasInitialValue,
+            ._hasStorage, ._nonoverride, ._private, ._show_in_interface, .dynamicCallable, .gkInspectable, .ibAction,
+            .ibOutlet
+        ])
+    #endif
+
+        // removed on Swift 4.2
+        expected.subtract([.objcNonLazyRealization, .inlineable, .versioned])
+        // removed o nSwift 4.1
```
Comment (jpsim) [inline Tests/SourceKittenFrameworkTests/SourceKitTests.swift]: Small nitpick:

```suggestion
        // removed on Swift 4.1
```

But also these should probably say "removed/added in Swift x.y" ("in" vs "on").

File: azure-pipelines.yml [hunk 10]
```diff
@@ -12,6 +12,12 @@ jobs:
         containerImage: norionomura/swiftlint:swift-4.2.0
       swift421:
         containerImage: norionomura/swiftlint:swift-4.2.1
+      swift422:
+        containerImage: norionomura/swiftlint:swift-4.2.2
+      swift423:
+        containerImage: norionomura/swiftlint:swift-4.2.3
+      swift50:
+        containerImage: norionomura/swift:5.0
```
Comment (jpsim) [inline azure-pipelines.yml]: Is there a reason to use this image over the official one?

```suggestion
        containerImage: swift:5.0
```

File: Source/SourceKittenFramework/Xcode.swift [hunk 11]
```diff
@@ -82,15 +103,7 @@ internal enum XcodeBuild {
                                            inPath: String) -> [XcodeBuildSetting]? {
         let arguments = xcodeBuildArguments + ["-showBuildSettings", "-json"]
         let outputData = XcodeBuild.launch(arguments: arguments, inPath: inPath, pipingStandardError: false)
-
-        let decoder = JSONDecoder()
-
-#if os(Linux) && !swift(>=4.2.1)
-        // Handled in `XcodeBuildSetting`.
-#else
-        decoder.keyDecodingStrategy = .convertFromSnakeCase
```
[the remark this answers is not in the document]
Comment (norio-nomura) [reply Source/SourceKittenFramework/Xcode.swift]: `.convertFromSnakeCase` does not work for keys on decoding `[String:String]` in Swift 5.
```terminal.sh-session
$ pbpaste
import Foundation

var json = """
{
    "action": "test",
    "buildSettings": {
        "SWIFT_MODULE_NAME": "sourcekitten"
    }
}
""".data(using: .utf8)!
struct XcodeSetting: Decodable {
    var action: String
    var buildSettings: [String: String]
}
let decoder = JSONDecoder()
decoder.keyDecodingStrategy = .convertFromSnakeCase
let xcodeSetting = try decoder.decode(XcodeSetting.self, from: json)
dump(xcodeSetting)
$ pbpaste|swift42 -
▿ main.XcodeSetting
  - action: "test"
  ▿ buildSettings: 1 key/value pair
    ▿ (2 elements)
      - key: "swiftModuleName"
      - value: "sourcekitten"
$ pbpaste|swift50 -
▿ main.XcodeSetting
  - action: "test"
  ▿ buildSettings: 1 key/value pair
    ▿ (2 elements)
      - key: "SWIFT_MODULE_NAME"
      - value: "sourcekitten"
```

[the remark this answers is not in the document]
Comment (jpsim) [reply Source/SourceKittenFramework/Xcode.swift]: Ah, shame. Do you know if this was intentional on the Swift team's part, or if it's a regression bug?

[the remark this answers is not in the document]
Comment (norio-nomura) [reply Source/SourceKittenFramework/Xcode.swift]: I think this is the intended behavior introduced in https://github.com/apple/swift/pull/16238.

File: Package.swift [hunk 12]
```diff
@@ -8,9 +8,9 @@ let package = Package(
         .library(name: "SourceKittenFramework", targets: ["SourceKittenFramework"])
     ],
     dependencies: [
-        .package(url: "https://github.com/Carthage/Commandant.git", from: "0.15.0"),
-        .package(url: "https://github.com/drmohundro/SWXMLHash.git", from: "4.7.5"),
-        .package(url: "https://github.com/jpsim/Yams.git", from: "1.0.1"),
+        .package(url: "https://github.com/Carthage/Commandant.git", .branch("master")),
+        .package(url: "https://github.com/drmohundro/SWXMLHash.git", .upToNextMinor(from: "4.8.0")),
```
[the remark this answers is not in the document]
Comment (norio-nomura) [reply Package.swift]: I do not think that the difference will be only `SWIFT_VERSION` even in future SWXMLHash 4.9.x.

File: Source/SourceKittenFramework/Xcode.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (jpsim) [reply Source/SourceKittenFramework/Xcode.swift]: ok. Oh well, it was nice while it lasted... 24 hours 😂

File: Package.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (jpsim) [reply Package.swift]: Right good call!

File: Tests/SourceKittenFrameworkTests/SourceKitTests.swift (hunk 7 above)
Comment (norio-nomura) [reply Tests/SourceKittenFrameworkTests/SourceKitTests.swift]: Certainly it would be better to refoctoring. But is it better to include them in this PR?

Comment (jpsim) [reply Tests/SourceKittenFrameworkTests/SourceKitTests.swift]: No it's ok, we can do it later.

Comment (norio-nomura) [reply Tests/SourceKittenFrameworkTests/SourceKitTests.swift]: #586

File: Source/SourceKittenFramework/SwiftDeclarationAttributeKind.swift (hunk 1 above)
Comment (norio-nomura) [reply Source/SourceKittenFramework/SwiftDeclarationAttributeKind.swift]: I'll drop them from `SwiftDeclarationAttributeKind`

**Comment by jpsim** [comment]
Only thing left I think is removing the duplicate SwiftDeclarationAttributeKind cases: https://github.com/jpsim/SourceKitten/pull/581#discussion_r272844757

Unless I'm not understanding why they're needed.

File: Source/SourceKittenFramework/SwiftDeclarationAttributeKind.swift (hunk 1 above)
Comment (norio-nomura) [reply Source/SourceKittenFramework/SwiftDeclarationAttributeKind.swift]: Removed on e43887a

**Comment by jpsim** [comment]
👍 LGTM. Feel free to merge when you're ready.

**Comment by norio-nomura** [comment]
Thanks! 🙏

**Comment by jpsim** [comment]
No, thank YOU! I should be able to cut a release tonight.

**Pull request merged**